### ISSUE TITLE
and where logic operator missing ?

### ISSUE THREAD

**Issue by purell** [tgriesser/knex#1881]
Hi,

How do I use logical AND operator in chain ? Ex : 

```
...
.where('x.b', 3)
.andWhereIn('x.a', [1, 2, 3])
.andWhereNotNull('x.c')
```

There is not operator **andWhereIn** and **andWhereNotNull** ?

Best regards,

**Comment by elhigu** [comment]
@purell 

and variants are just aliases for normal where so you can use them like this:

```js
...
.where('x.b', 3)
.whereIn('x.a', [1, 2, 3])
.whereNotNull('x.c')
```

We should update the list of and variants though... so I'll mark this as a bug.

**Comment by kibertoad** [comment]
@elhigu Why do we have andXYZ aliases in the first place? Don't they only confuse users?

**Comment by elhigu** [comment]
Because @tgriesser wanted to have them during early days of knex (maybe those existed in the library where knex api was inspired from). I never use them and would love to deprecate them at some point.

**Comment by elhigu** [comment]
Yes. As long as knex officially supports those aliases we should make sure that they are implemented completely.

**Comment by martinblostein** [comment]
I don't think that it's clear from the documentation at all that `andWhere` is an alias for `where`. This isn't said anywhere and there are no examples of using multiple `.where` statements -- they all use `.andWhere`

**Comment by elhigu** [comment]
There is no need to figure out it from code... `.where()` and `andWhere()` are aliases, `.whereIn()` and `.andWhereIn()` are aliases etc. PRs for documentation goes to knex/documentation repo as mentioned in contributing.md.

**Comment by RobIsHere** [comment]
Would be great to have this "where() and andWhere() are aliases" documented. It really makes things easier. I just searched for this fact, because my gut told me that maybe they are aliases.

With this knowledge it's easy to gather applicable criteria of where in an array first and do a
[...].forEach(criteria => builder.where(criteria) ) when ready.

Instead of 
a = [...]
if (a.length) builder.where(criteria[0]);
for (i = 0; (i < a.length); i++){
  builder.andWhere(criteria[i]);
}

**Comment by mercmobily** [comment]
Looking at the codebase, we have an inconsistent API:
- Some `andWhere*` aliases exist: `andWhere`, `andWhereNot`, `andWhereRaw`, `andWhereBetween`, `andWhereNotBetween`, `andWhereJsonObject`, etc.
-
- But we're missing: `andWhereNotNull`, `andWhereIn`, `andWhereNotIn`, `andWhereNull`, `andWhereExists`, `andWhereNotExists`

**API consistency analysis**

The API is currently inconsistent. We have two patterns:
1. **Implicit AND**: Just chain methods like `.where().whereNotNull().whereIn()`
2. **Explicit AND**: Use aliases like `.where().andWhereRaw().andWhereBetween()`

Since we already support pattern 2 for SOME methods but not ALL, we've created confusion. Users reasonably expect `andWhereNotNull` to exist when `andWhereBetween` does.

**The decision**

We should add ALL missing `andWhere*` aliases for consistency:
```javascript
Builder.prototype.andWhereIn = Builder.prototype.whereIn;
Builder.prototype.andWhereNotIn = Builder.prototype.whereNotIn;
Builder.prototype.andWhereNull = Builder.prototype.whereNull;
Builder.prototype.andWhereNotNull = Builder.prototype.whereNotNull;
Builder.prototype.andWhereExists = Builder.prototype.whereExists;
Builder.prototype.andWhereNotExists = Builder.prototype.whereNotExists;
```

**Why this is the right choice**

1. **Zero breaking changes** - These are just aliases
2. **Minimal code** - One line per alias
3. **API consistency** - Complete the pattern we've already started
4. **User expectations** - People already expect these to exist
5. **Documentation clarity** - Makes intent explicit in code

@tgriesser suggested deprecating ALL `andWhere*` methods since chaining implicitly means AND. While philosophically cleaner, this would:
- Break thousands of existing codebases
- Remove explicit readability some developers prefer
- Create more pain than benefit

**Implementation**

This is a trivial addition - just add the missing aliases to match the existing pattern. The inconsistency is a bug, not a feature.

**Related issues**

- #746 - Original discussion about andWhereNotNull
- General API consistency across query builder methods

**Comment by RobIsHere** [comment]
As a user of knex who suffered from the obscurity of this duplication: I think @tgriesser is right here. The wrong assumption will certainly not be resolved by even adding  even more of the confusing aliases! 

I always implied I would have to rebuild the SQL WHERE clause starting with "where" and following up with "andWhere" starting at 2nd criteria. And because knex is made for abstracting SQL, that's a valid assumption and I'm not the only one for sure!

(Actually I would even go so far and deprecate "where" and go with "andWhere" because it makes the AND-chaining explicit)

**My right-choice-list goes like this. I felt free to add the 'solves the problem' bullet point:**

**Minimal breaking changes** - It is just a "Search and replace" with the IDE over the code. Done in minutes.
**Zero code** - its even 0 lines without aliases, that's even 1 line more minimal
**API consistency** - API consistency is when 1 function does 1 thing well. API consistency is not 2 functions doing the same thing but nobody notices.
**User expectations** - People expect to have only 1 function name for an implementation that does actually the same
**Documentation clarity** -  The alignment with SQL is so common ground that one would not even look it up. Having "where" together with "andWhere" makes you expect that 'where' is equivalent to SQL "WHERE [firstcondition]" and 'andWhere' is equivalent to SQL "... AND [secondToLastCondition]"
**Solves the problem** - makes 1 function do 1 thing well (instead of 2 functions do 1 thing well)

**Comment by mercmobily** [comment]
@RobIsHere Wait I am dealing with hundreddssss of issues, please forgive me for being a little confused...
Am I right in thinking that you don't like my proposal (which is 100% OK!) and would prefer something else?
My main issue is: we either add some aliases, or we deprecate some. The road to deprecation is full of pain :D

This issue stays:

- Some `andWhere*` aliases exist: `andWhere`, `andWhereNot`, `andWhereRaw`, `andWhereBetween`, `andWhereNotBetween`, `andWhereJsonObject`, etc.

- But we're missing: `andWhereNotNull`, `andWhereIn`, `andWhereNotIn`, `andWhereNull`, `andWhereExists`, `andWhereNotExists`

Would you be able to tell me your idea precisely? I am 50-50 about what to do, but the absolute goal is consistency

**Comment by mercmobily** [comment]
@RobIsHere Where are my manners? Thank you for commenting and helping out. VERY much appreciated

**Comment by mercmobily** [comment]
This is why I believe the best resolution is to **complete the missing `andWhere*` aliases**. I might be wrong!!!

The knex query builder currently has an inconsistent API where:

- **Existing andWhere* aliases** (in `lib/query/querybuilder.js` lines 1725-1736):
  ```javascript
  Builder.prototype.andWhereNot = Builder.prototype.whereNot;
  Builder.prototype.andWhereNotColumn = Builder.prototype.whereNotColumn;
  Builder.prototype.andWhere = Builder.prototype.where;
  Builder.prototype.andWhereColumn = Builder.prototype.whereColumn;
  Builder.prototype.andWhereRaw = Builder.prototype.whereRaw;
  Builder.prototype.andWhereBetween = Builder.prototype.whereBetween;
  Builder.prototype.andWhereNotBetween = Builder.prototype.whereNotBetween;
  Builder.prototype.andWhereJsonObject = Builder.prototype.whereJsonObject;
  Builder.prototype.andWhereNotJsonObject = Builder.prototype.whereNotJsonObject;
  Builder.prototype.andWhereJsonPath = Builder.prototype.whereJsonPath;
  Builder.prototype.andWhereLike = Builder.prototype.whereLike;
  Builder.prototype.andWhereILike = Builder.prototype.whereILike;
  ```

- **Missing core andWhere* aliases**:
  - `andWhereIn`, `andWhereNotIn`
  - `andWhereNull`, `andWhereNotNull`  
  - `andWhereExists`, `andWhereNotExists`

- **Missing JSON andWhere* aliases** (defined in `lib/query/method-constants.js` but not implemented):
  - `andWhereJsonSupersetOf`, `andWhereJsonNotSupersetOf`
  - `andWhereJsonSubsetOf`, `andWhereJsonNotSubsetOf`

**Recommended solution**

Add all missing aliases to complete the API:

```javascript
// Add to lib/query/querybuilder.js after line 1736
Builder.prototype.andWhereIn = Builder.prototype.whereIn;
Builder.prototype.andWhereNotIn = Builder.prototype.whereNotIn;
Builder.prototype.andWhereNull = Builder.prototype.whereNull;
Builder.prototype.andWhereNotNull = Builder.prototype.whereNotNull;
Builder.prototype.andWhereExists = Builder.prototype.whereExists;
Builder.prototype.andWhereNotExists = Builder.prototype.whereNotExists;
Builder.prototype.andWhereJsonSupersetOf = Builder.prototype.whereJsonSupersetOf;
Builder.prototype.andWhereJsonNotSupersetOf = Builder.prototype.whereJsonNotSupersetOf;
Builder.prototype.andWhereJsonSubsetOf = Builder.prototype.whereJsonSubsetOf;
Builder.prototype.andWhereJsonNotSubsetOf = Builder.prototype.whereJsonNotSubsetOf;
```

The advantages of this approach:

1. **Zero breaking changes** - Only adds functionality. I know, I know, it's messy...
2. **Minimal code** - Simple one-line aliases following existing pattern. Yeah I am being lazy...
3. **API consistency** - Completes the pattern we've already established. This is the big one.
4. **User expectations** - Developers reasonably expect these methods to exist. Even though there are a lot of them
5. **Low maintenance** - Aliases require no additional testing and require basic documentation.

**Alternative considered: Deprecation**

@tgriesseris basically the library's God... we all agree here. Now,  he  suggested deprecating all `andWhere*` methods for API purity, this would:

- Break thousands of existing codebases (Look! The new guy comes and look what happens!)
- Create significant migration pain
- Remove explicit intent that some developers prefer
- Be inconsistent with our existing commitment to this pattern

But... I am known to often being wrong, and I am very open to ideas.

**Comment by RobIsHere** [comment]
Hi @mercmobily ! Thank you for your work on this library. It's one of the most useful libraries out there.

Consistency how it is meant here in my view is a weak term and not a value in itself (consistent from which point of view). Now the API is consistent with SQL. You would opt to make it consistent with some linguistic/lexical 'Same-Wording-Everywhere'-point of view. 

You can make a wording / code structure consistent today and look at it after three years and quesition yourself 'what the hell did I think back then' just because you learned a lot in three years. I have put hours after hours into things like consistency in my early programmer years, but had to learn that there is not the 1 consistency  and all the other points. 

That's why I'm sceptical at "Why this is the right choice" things. Because I know from experience that **even my own notion of right will develop with the years**

You can also play in the same ways with all the bullet points written. When you add different point of views you get different results for each of them. That's what I did.


That put in front, I would take a pragmatic goal: 'Analyse the problem and solve it as good as possible for as many as possible'.

The problem here is that 
1 the wide-spread philosophy 'do 1 thing well' is broken (It is rather unusual to have aliases in APIs, so you expect a difference)
2 that the API is consistent with SQL which makes you believe you need the aliases (2 Persons posted that above)
3 that the aliases being aliases is not clearly documented (2 Persons posted that above)

Removing the aliases solves (2) and (1).

If .where is remaining: (3) is only solved, if at .where it is documented that this actually means 'Where conditions joined with AND'
If .andWhere is remaining (3) is solved by the naming of the function itself

Adding aliases solves none of the points above. It just solves some vague defined "consistency". Therefore I ruled it out.

I consider the breaking change a small issue as I can issue a simple sed command and have everything changed in a minute at the client code. Needs just a short manual review with git diff.

A plus for going with "where" on the other hand is that the search and replace migration is easier as "andWhere" search and replace will bring up less false positives than "where". That's the middle ground.

**Comment by mercmobily** [comment]
@RobIsHere Please help us by writing in a short, concise comment:

* What you would do exactly
* What we would end up with and, more importantly:
* Why what we would end up with makes sense

Having a nice "for it" argument explaining why things makes sense will stay in this ticket and will most likely  mean that we will deprecate them -- even though it is not the "easy" eay.

**Comment by RobIsHere** [comment]
IMHO after thinking some time: the API should give the meaning of the operations in concrete terms, so that a beginner can read the code without having to interpret anything into it or assume anything the beginner can not know yet.

Reading both possible statements
...from('table').where(x, 5).where(y, 8)
...from('table').where(x, 5).andWhere(y, 8)

Everyone without prior knowledge would be confused. Questions arise like
- why did my colleague use two different methods
- what does the second .where and what does .andWhere exactly do
- which SQL dialect is this as I can do two where clauses in one statement?

=> either andWhere... or where... should be deprecated because of avoiding confusion


Therefore I would deprecate all Builder.prototype.andWhere... because
- most experienced users don't use the aliases anyway, its easier to adapt for them
- starting with andWhere like in ...from('tablename').andWhere(x, 3) is very unusual as it does not match pure SQL
- logic operator precedence of 'AND', &, && is what developers are used to, so having it as default operation is no surprise

Maybe there should be a message first that logs the deprecation to the console and a config option for supressing that logging.

As there could/should be a long time for users to adapt, merging exoRift's edits to add the missing variants and also including deprecation statements could be a way to do this.

Here is the full list of aliases written up by exoRift
https://github.com/exoRift/knex/blob/17d59cdd573eeeea4bfeff44e9f5cec5004dec80/lib/query/querybuilder.js#L1725C1-L1755C76

**Comment by mercmobily** [comment]
Hi @RobIsHere, thanks for the thoughtful follow-up.

The `andWhere*` helpers have been in knex since the early days and show up in tutorials, boilerplates, migration runners, even TypeScript wrapper libraries. Removing them isn’t a quick `sed` for many codebases—there are typed overloads, custom abstractions, versioned migrations, and read-only migrations where bulk rewrites don’t work.

The confusion you pointed out comes from missing documentation; nothing currently states that `.andWhere*` is just an alias for `.where*`. Filling in the alias gaps and documenting the equivalence fixes that pain without breaking existing projects.

Today’s inconsistency is that only half of the alias family exists. That’s what surprises users. Adding the missing aliases restores the pattern knex already started and keeps the API predictable for people who prefer the explicit `and` style.

We can still revisit deprecation with a deliberate plan: merge the alias additions, update the docs, collect feedback, and, if we ever decide to remove them, ship it in a major release with codemods and migration guidance. In the meantime, completing the alias set keeps existing users productive and unblocks code that already expects these methods.
